`TypeError` on SerialAdapter `read_bytes(-1)`

When I use `self.read_bytes(-1)` from my instrument using a `SerialAdapter`, I get the error:

> C:\Program Files\Python310\lib\ctypes\__init__.py:66: TypeError

With the traceback:

```
..\..\..\pymeasure\instruments\velleman\velleman_k8090.py:247: in read
    response = self.read_bytes(-1)
..\..\..\pymeasure\instruments\instrument.py:153: in read_bytes
    return self.adapter.read_bytes(count, **kwargs)
..\..\..\pymeasure\adapters\adapter.py:123: in read_bytes
    read = self._read_bytes(count, break_on_termchar, **kwargs)
..\..\..\pymeasure\adapters\serial.py:108: in _read_bytes
    return self.connection.read(1e99 if count == -1 else count, **kwargs)
..\..\..\venv\lib\site-packages\serial\serialwin32.py:278: in read
    buf = ctypes.create_string_buffer(n)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

init = 1e+99, size = None

    def create_string_buffer(init, size=None):
        """create_string_buffer(aBytes) -> character array
        create_string_buffer(anInteger) -> character array
        create_string_buffer(aBytes, anInteger) -> character array
        """
        if isinstance(init, bytes):
            if size is None:
                size = len(init)+1
            _sys.audit("ctypes.create_string_buffer", init, size)
            buftype = c_char * size
            buf = buftype()
            buf.value = init
            return buf
        elif isinstance(init, int):
            _sys.audit("ctypes.create_string_buffer", None, init)
            buftype = c_char * init
            buf = buftype()
            return buf
>       raise TypeError(init)
E       TypeError: 1e+99

C:\Program Files\Python310\lib\ctypes\__init__.py:66: TypeError

```

It looks like reading 1e99 bytes no (longer?) works.

--- a/pymeasure/adapters/serial.py
+++ b/pymeasure/adapters/serial.py
@@ -109,12 +109,14 @@ class SerialAdapter(Adapter):
             # For -1 we empty the buffer completely
             return self._read_bytes_until_timeout()
 
-    def _read_bytes_until_timeout(self, **kwargs):
-        """Read from the serial until a timeout occurs, regardless of how many bytes."""
-        # `Serial.readlines()` sounds appropriate instead but it turns out it has an
-        # unpredictable timeout
+    def _read_bytes_until_timeout(self, chunk_size=256, **kwargs):
+        """Read from the serial until a timeout occurs, regardless of the number of bytes.
+
+        :chunk_size: The number of bytes attempted to in a single transaction.
+            Multiple of these transactions will occur.
+        """
+        # `Serial.readlines()` has an unpredictable timeout, see PR #866
         data = bytes()
-        chunk_size = 256
         while True:
             chunk = self.connection.read(chunk_size, **kwargs)
             data += chunk

--- a/pymeasure/adapters/serial.py
+++ b/pymeasure/adapters/serial.py
@@ -94,7 +94,7 @@ class SerialAdapter(Adapter):
         """Read a certain number of bytes from the instrument.
 
         :param int count: Number of bytes to read. A value of -1 indicates to
-            read from the whole read buffer.
+            read from the whole read buffer (waits for timeout).
         :param bool break_on_termchar: Stop reading at a termination character.
         :param \\**kwargs: Keyword arguments for the connection itself.
         :returns bytes: Bytes response of the instrument (including termination).
@@ -103,9 +103,12 @@ class SerialAdapter(Adapter):
             return self.connection.read_until(self.read_termination.encode(),
                                               count if count > 0 else None,
                                               **kwargs)
+        elif count >= 0:
+            return self.connection.read(count, **kwargs)
         else:
-            # At -1 we read a very large number of bytes, which can be considered the whole buffer.
-            return self.connection.read(1e99 if count == -1 else count, **kwargs)
+            # For -1 we empty the buffer completely
+            num_bytes = self.connection.in_waiting + 100  # Read what's already queued + some extra
+            return self.connection.read(num_bytes, **kwargs)
 
     def flush_read_buffer(self):
         """Flush and discard the input buffer."""

--- a/pymeasure/adapters/serial.py
+++ b/pymeasure/adapters/serial.py
@@ -107,8 +107,19 @@ class SerialAdapter(Adapter):
             return self.connection.read(count, **kwargs)
         else:
             # For -1 we empty the buffer completely
-            num_bytes = self.connection.in_waiting + 100  # Read what's already queued + some extra
-            return self.connection.read(num_bytes, **kwargs)
+            return self._read_bytes_until_timeout()
+
+    def _read_bytes_until_timeout(self, **kwargs):
+        """Read from the serial until a timeout occurs, regardless of how many bytes."""
+        # `Serial.readlines()` sounds appropriate instead but it turns out it has an
+        # unpredictable timeout
+        data = bytes()
+        chunk_size = 256
+        while True:
+            chunk = self.connection.read(chunk_size, **kwargs)
+            data += chunk
+            if len(chunk) < chunk_size:  # If fewer bytes got returned, we had a timeout
+                return data
 
     def flush_read_buffer(self):
         """Flush and discard the input buffer."""

--- a/tests/adapters/test_serial.py
+++ b/tests/adapters/test_serial.py
@@ -73,6 +73,12 @@ def test_read_bytes_unlimited(adapter):
     assert adapter.read_bytes(-1) == b"basd\x02\nfasdf\n"
 
 
+def test_read_bytes_unlimited_long(adapter):
+    """Test whether all bytes are returned when a lot of data is sent."""
+    adapter.write_bytes(b"abcde" * 50)
+    assert adapter.read_bytes(-1) == b"abcde" * 50
+
+
 @pytest.mark.parametrize("count", (-1, 8))
 def test_read_bytes_break_on_termchar(adapter, count):
     adapter.read_termination = "\n"

--- a/tests/adapters/test_serial_with_loopback.py
+++ b/tests/adapters/test_serial_with_loopback.py
@@ -1,0 +1,80 @@
+#
+# Test the serial adapter using looped COM ports.
+# This looping needs to be done outside this script. One suggestion is
+# using the application `com0com`.
+#
+
+import pytest
+
+from serial import Serial
+from time import time
+from pymeasure.adapters import SerialAdapter
+
+
+ADAPTER_TIMEOUT = 3.0  # Seconds
+
+
+@pytest.fixture()
+def adapter(connected_device_address):
+    """Call with `--device-address="COM3,COM4"` to select two devices."""
+    device = connected_device_address.split(",")[0]
+    return SerialAdapter(
+        device,
+        baudrate=19200,
+        timeout=ADAPTER_TIMEOUT,
+        read_termination=chr(0x0F),
+    )
+
+
+@pytest.fixture()
+def loopback(connected_device_address):
+    """See `adapter()`."""
+    device = connected_device_address.split(",")[1]
+    return Serial(device, baudrate=19200)
+
+
+def test_read(adapter, loopback):
+    """Regular read with fixed number of bytes."""
+    loopback.write(b"abc")
+    result = adapter.read_bytes(3)
+
+    assert len(result) == 3
+
+
+def test_read_all_short(adapter, loopback):
+    """Read with unlimited number of bytes"""
+    loopback.write(b"a")
+    start = time()
+    result = adapter.read_bytes(-1)
+    elapsed = time() - start
+
+    assert len(result) == 1
+    assert ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
+
+
+def test_read_all(adapter, loopback):
+    loopback.write(b"aaaaabbbbbccccceeeee" * 50)
+    start = time()
+    result = adapter.read_bytes(-1)
+    elapsed = time() - start
+
+    assert len(result) == 20 * 50
+    assert ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
+
+
+def test_read_all_exact_chunck_size(adapter, loopback):
+    loopback.write(b"a" * 256)
+    start = time()
+    result = adapter.read_bytes(-1)
+    elapsed = time() - start
+
+    assert len(result) == 256
+    assert ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
+
+
+def test_read_all_with_newline(adapter, loopback):
+    data = b"aaaaabbbbb\nccccceeeee" * 50
+    loopback.write(data)
+    result = adapter.read_bytes(-1)
+
+    assert len(result) == 21 * 50

--- a/tests/adapters/test_serial_with_loopback.py
+++ b/tests/adapters/test_serial_with_loopback.py
@@ -1,9 +1,3 @@
-#
-# Test the serial adapter using looped COM ports.
-# This looping needs to be done outside this script. One suggestion is
-# using the application `com0com`.
-#
-
 import pytest
 
 from serial import Serial
@@ -11,70 +5,64 @@ from time import time
 from pymeasure.adapters import SerialAdapter
 
 
-ADAPTER_TIMEOUT = 3.0  # Seconds
+class TestSerialLoopback:
+    """Test serial adapter with two real but looped COM ports.
 
+    A pair of COM ports that are looped will read the written data from the other.
+    This looping needs to be done outside this script. One suggestion is using the application
+    `com0com <https://com0com.sourceforge.net/>`_.
+    Alternatively two UART adapters could be used, that have their RX/TX physically connected.
 
-@pytest.fixture()
-def adapter(connected_device_address):
-    """Call with `--device-address="COM3,COM4"` to select two devices."""
-    device = connected_device_address.split(",")[0]
-    return SerialAdapter(
-        device,
-        baudrate=19200,
-        timeout=ADAPTER_TIMEOUT,
-        read_termination=chr(0x0F),
-    )
+    Call PyTest with the argument``--device-address="COM1,COM2"`` to specify the serial addresses,
+    minding the comma for the separation of the two ports.
+    """
 
+    ADAPTER_TIMEOUT = 1.0  # Seconds
 
-@pytest.fixture()
-def loopback(connected_device_address):
-    """See `adapter()`."""
-    device = connected_device_address.split(",")[1]
-    return Serial(device, baudrate=19200)
+    @pytest.fixture()
+    def adapter(self, connected_device_address):
+        device = connected_device_address.split(",")[0]
+        return SerialAdapter(
+            device,
+            baudrate=19200,
+            timeout=self.ADAPTER_TIMEOUT,
+            read_termination=chr(0x0F),
+        )
 
+    @pytest.fixture()
+    def loopback(self, connected_device_address):
+        """See `adapter()`."""
+        device = connected_device_address.split(",")[1]
+        return Serial(device, baudrate=19200)
 
-def test_read(adapter, loopback):
-    """Regular read with fixed number of bytes."""
-    loopback.write(b"abc")
-    result = adapter.read_bytes(3)
+    def test_read(self, adapter, loopback):
+        """Regular read with fixed number of bytes."""
+        loopback.write(b"abc")
+        result = adapter.read_bytes(3)
 
-    assert len(result) == 3
+        assert len(result) == 3
 
+    @pytest.mark.parametrize("data", [
+        b"a",                               # Short data
+        b"aaaaabbbbbccccceeeee" * 50,       # A lot data
+        b"a" * 256,                         # Exactly one chunk size
+        b"aaaaabbbbb\nccccceeeee" * 50,     # With a newline (should be ignored)
+    ])
+    def test_read_all(self, adapter, loopback, data):
+        """Read with undefined number of bytes - also confirm timeout duration is correct."""
+        loopback.write(data)
+        start = time()
+        result = adapter.read_bytes(-1)
+        elapsed = time() - start
 
-def test_read_all_short(adapter, loopback):
-    """Read with unlimited number of bytes"""
-    loopback.write(b"a")
-    start = time()
-    result = adapter.read_bytes(-1)
-    elapsed = time() - start
+        assert result == data
+        assert self.ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
 
-    assert len(result) == 1
-    assert ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
+    @pytest.mark.parametrize("chunk", [1, 256, 10000])
+    def test_read_varied_chunk_size(self, adapter, loopback, chunk):
+        """Read with undefined number of bytes with non-default chunk size."""
+        data = b"abcde" * 10
+        loopback.write(data)
+        result = adapter.read_bytes(-1)
 
-
-def test_read_all(adapter, loopback):
-    loopback.write(b"aaaaabbbbbccccceeeee" * 50)
-    start = time()
-    result = adapter.read_bytes(-1)
-    elapsed = time() - start
-
-    assert len(result) == 20 * 50
-    assert ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
-
-
-def test_read_all_exact_chunck_size(adapter, loopback):
-    loopback.write(b"a" * 256)
-    start = time()
-    result = adapter.read_bytes(-1)
-    elapsed = time() - start
-
-    assert len(result) == 256
-    assert ADAPTER_TIMEOUT == pytest.approx(elapsed, abs=0.1)
-
-
-def test_read_all_with_newline(adapter, loopback):
-    data = b"aaaaabbbbb\nccccceeeee" * 50
-    loopback.write(data)
-    result = adapter.read_bytes(-1)
-
-    assert len(result) == 21 * 50
+        assert result == data

